File: leetcode/826. Most Profit Assigning Work/826.java
class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->(a[0]-b[0]));
        int N = profit.length, best = 0;
        Arrays.sort(worker);
        int res = 0;
        for(int i=0; i<N; i++){
            pq.add(new int[]{difficulty[i], profit[i]});
        }

        for(int w:worker){
            while(!pq.isEmpty() && w >= pq.peek()[0]){
                best = Math.max(best, pq.peek()[1]);
                pq.poll();
            }
            res += best;
        }


        return res;
    }
}


// TC: O(Dlog(D) + Wlog(W))
// SC: O(D)

File: leetcode/Medium/826. Most Profit Assigning Work/826.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>> tasks;
        int N = difficulty.size(), res = 0, best = 0;

        for(int i=0; i<N; i++){
            tasks.push_back(make_pair(difficulty[i], profit[i]));
        }

        sort(tasks.begin(), tasks.end());
        sort(worker.begin(), worker.end());

        int i=0;
        for(int &w:worker){
            while(i<N && w>=tasks[i].first)
                best = max(best, tasks[i++].second);
            res += best;
        }
        return res;
    }
};

// TC: O(Dlog(D) + Wlog(W)) 
// SC: O(D)